File: typescript/queen-attack/queen-attack.ts
type Position = readonly [number, number]
type Queens = {[key: string]: Position}


class Queen {
  public row: number
  public col: number

  constructor([row, col]: Position) {
    if (row < 0 || row > 7 || col < 0 || col > 7)
      throw new Error('Queen must be placed on the board')

    this.row = row
    this.col = col
  }

  get position(): Position {
    return [this.row, this.col]
  }

  equals(other: Queen) {
    return this.row === other.row && this.col === other.col
  }
}


export class QueenAttack {
  private b: Queen
  private w: Queen

  constructor(queens?: Queens) {
    queens ??= {}
    queens.black ??= [0, 3]
    queens.white ??= [7, 3]

    this.b = new Queen(queens.black)
    this.w = new Queen(queens.white)

    if (this.b.equals(this.w))
      throw new Error('Queens cannot share the same space')
  }

  get black() { return this.b.position; }
  get white() { return this.w.position; }

  get canAttack(): boolean {
    const dx = Math.abs( this.b.row - this.w.row )
    const dy = Math.abs( this.b.col - this.w.col )
    return (dx === 0 || dy === 0 || dx === dy)
  }

  toString(): string {
    const board = Array.from({length: 8}, () => new Array(8).fill('_'))
    board[ this.b.row ][ this.b.col ] = 'B'
    board[ this.w.row ][ this.w.col ] = 'W'
    return board.map((row) => row.join(' ')).join('\n')
  }
}

File: nim/queen-attack/queen_attack.nim
const Size = 8    # board is Size x Size

type
  Queen* = ref object
    row*: int
    col*: int


proc initQueen*(row: int, col: int): Queen =
  if row < 0:
    raise newException(ValueError, "queen must have positive row")
  if col < 0:
    raise newException(ValueError, "queen must have positive column")
  if row >= Size:
    raise newException(ValueError, "queen must have row on board")
  if col >= Size:
    raise newException(ValueError, "queen must have column on board")
  Queen(row: row, col: col)


proc `==`*(q1: Queen, q2: Queen): bool =
  q1.row == q2.row and q1.col == q2.col


proc checkPositions(q1: Queen, q2: Queen): void =
  if q1 == q2:
    raise newException(ValueError, "Queens cannot occupy same position")


proc canAttack*(q1: Queen, q2: Queen): bool =
  checkPositions q1, q2
  q1.row == q2.row or
  q1.col == q2.col or
  (q1.row - q2.row).abs == (q1.col - q2.col).abs


proc cell(x:int, y:int, q1: Queen, q2: Queen): char =
  var pos = Queen(row: x, col: y)
  if pos == q1: return 'W'
  if pos == q2: return 'B'
  return '_'


proc board*(q1: Queen, q2: Queen): string =
  checkPositions q1, q2
  for x in 0 ..< Size:
    for y in 0 ..< Size:
      result &= cell(x, y, q1, q2)
    result &= "\n"
